### PR TITLE
release 5.1.0

### LINKED ISSUE #1742: [BUG] @prizm-ui/components - Некорректные отступы у tree
### Библиотека

- [ ] `@prizm-ui/core`
- [x] `@prizm-ui/components`
- [ ] `@prizm-ui/install`
- [ ] `@prizm-ui/icons`
- [ ] `@prizm-ui/flag-icons`
- [ ] `@prizm-ui/theme`
- [ ] `@prizm-ui/charts`
- [ ] `@prizm-ui/ast`
- [ ] `@prizm-ui/nx-plugin`
-

### Компонент

`PrizmTree`

### Версия

Операционная система:

`Linux 6.1.41-1-MANJARO`

Браузер:

`24.1.5.794 stable (64-bit)`

NodeJS:

```
➜ node --version
v16.20.1
```
### Dependencies:
```
prizm v.1.16.0 (но и на последних воспроизводится)
```

### Проект в котором используется Prizm

ZQL

### Описание проблемы

Отступы не по [дизайну](https://www.figma.com/design/2KJZq6hHEdxQmFMRVTJWKT/%F0%9F%92%8EPrizm.-%D0%A1omponent-base-4.0?node-id=692-102190&t=wn4bhk0D2fWmazun-11) 

image

### PR DESCRIPTION
**Pull request by ZurabDev**
### Features
- feat(components/input-tree-multi-select): added ability to control checkbox position #2094
- feat(components/input-tree-select): added ability to control placeholder in search #2092
- feat(components/input-tree-multi-select): added ability to control placeholder in search #2092
- feat(components/input-tree-select): added autofocus for search input #2090
- feat(components/input-tree-multi-select): added autofocus for search input #2090
- feat(helpers/hint-overflow): added input to reserve extra space
- feat(components/file-upload): added input: files && added outputs: fileAdded, fileRemoved #1832
- feat(components/file-upload): added custom actions && outputs: actionEvent, afterFilesChange #1832
- feat(components/tree-select): added new component #2069
- feat(components/tree-multi-select): added new component #2069
- feat(components/switcher): extended switcher for pass as projection children #2069
- feat(doc): add link to github issue page
- feat(components/input-select): fixed position of search input for dropdown #1981 Note, there are layout changes: if you overridden the --prizm-select-item-padding variable in the project, there may be a regression.
- feat(components/input-multiselect): fixed position of search input for dropdown #1981 Note, there are layout changes: if you overridden the --prizm-select-item-padding variable in the project, there may be a regression.
- feat(components/checkbox): add box shadow for disabled radio and checkbox #2024
- feat(doc/multiselect): add example for async items in multiselect #1754
- feat(helpers/overflow): new directives host, item to hide overflowed items
- feat(helpers/context): new directives and pipes to pass context and get context
- feat(helpers/resize-observer): new util function to get observable of resize observer
- feat(helpers/map): map with ability to get changes stream
- feat(components/input-date-time): add time restrictions for min max values #1570
- feat(components/input-date-time-range): add time restrictions for min max values #1570
- feat(components/tree): added css variable for tree item height
- feat(components/paginator): add left direction support #1154
- feat(components/input-select): added input to control auto reposition dropdown window #1688
- feat(components/input-multi-select): added input to control auto reposition dropdown window
- feat(charts/area): replace PrizmChartsAriaOptions(deprecated now) by PrizmChartsAreaOptions
- feat(charts): add color input for area, line, column, pie, radar, radial bar, scatter, treemap, waterfall #1816
- feat(doc/charts): for line and column charts added ecamples with user's colors #1817
- feat(doc/icons): example how to use custom icon #1841
- feat(components/paginator): update paginator dictionary #1840 BREAKING CHANGE in dictionary - BREAKING CHANGE in dictionaries, why we do this read here
- feat(doc/slider): add form control for slider #531
- feat(components/file-upload): add hint for browse button when its disable due max files count reached #1769 - BREAKING CHANGE in dictionaries, why we do this read [here](https://github.com/zyfra/Prizm/discussions/1617)
- feat(components/file-upload): add translations for file size unit #1789 - BREAKING CHANGE in dictionaries, why we do this read [here](https://github.com/zyfra/Prizm/discussions/1617)
- feat(chore): update mr checklist template
- feat(components/accordion): border-bottom can be dded to accordion item header #1322
- feat(components/input-dates): added native date transformation for min max properties #1315
- feat(components/input-dates): added date string transformation for min max properties #1573
- feat(components/cron): added native date transformation for min max properties in date inputs #1315
- feat(components/input-layout-date): added new provider for ISO and UTC strings transformer for input layout date #1574
- feat(doc): remove examples for deleted prizm-input-date component
- feat(nx-mv): support up version with templates

### Bug fixes
- fix(helpers/hint-overflow): re calculate position after change window size #2093 #1983  #2052
- fix(components/chipd): re calculate position after change window size #2093
- fix(doc/version): add v5 version link to v4 version manager #2035
- fix(component/chips): chips overflows the border #1975
- fix(components/table): table sort cursor correction #1986
- fix(components/table): sorter count div should not reserve space when empty #2021
- fix(components/tabs): fix Cannot read properties of undefined (reading 'disconnect') #2013
- fix(components): added icons to peerDependencies #1989
- fix(components/button): correct outline buttons height #1399
- fix(components/navigation-menu): fix navigation item height #1964
- fix(doc): missing fix #1742 info added to 4.3.6 version in changelog
- fix(components/slider): add markForCheck for setinng position from value for form control correct work #531
- fix(components/accordion): accordion item header height should be 48px
- fix(components/tabs): view not updated for activeTabindexChage when tabs overflows #1863
- fix(components/tabs): tabindex should recalculate on close from dropdown #1948
- fix(components/listing-item): disabled items should not fire mouse events #1947

### Refactor
- refactor(components/confirm-popup): refactored directive, replaced old code to host directive

### Checklist:

- [x] Изменить версии в соответствующем массиве на ветке V4
  - apps/doc/src/app/version-manager/versions.constants.ts
- [ ] Изменить version на текущий в файле nxmv.json и запустив команду `npx nx generate nx-mv:apply -n v18 --var-version 5.2.0` важно передавать --var-version нужную версию
- [x] Добавить в каждый тип ваши изменения в этом PR
- [x] Удалите не нужные группы изменений для этого PR (Features, Bug Fixes, Breacking Changes, Refactor)
- [x] Добавить изменения в changelog о новой версии
- [ ] Составить readMe файл с описанием изменений на русском

### CODE REVIEW

**Comment by alexhawkins94** [comment]
FileUpload
Текстовка "Ожидание загрузки" стала переносится на новую строку.
chrome_qCRDVtizer

Раньше было в одну строку, нужно вернуть как было
chrome_FNIrxEout9

**Comment by alexhawkins94** [comment]
По поводу https://github.com/zyfra/Prizm/issues/2093  

При увеличении ширины инпута чипсы не разворачиваются автоматически из-под многоточия.
Пересчет и отрисовка свернутых чипсов происходит когда пользователь совершает клик по инпуту.

**Comment by alexhawkins94** [comment]
Сделать информацию об используемых иконок в отдельной вкладке.
Текущее отображение информации об иконках выглядит банально страшно.
image

**Comment by alexhawkins94** [comment]
Multiselect 
Ошибка в консоли на странице с примерами 
chrome_4sqo2RxIex

**Comment by alexhawkins94** [comment]
Multiselect
В outer мультиселекте чипсы стали выходить за нижнюю границу инпута.

image
image

**Comment by alexhawkins94** [comment]
Multiselect 
В примерах with object, transformer не отображаются названия выбранных айтемов в чипсах.
image

**Comment by alexhawkins94** [comment]
Вновь замечено большое потребление памяти вкладки в браузере с открытой витриной - порядка 830 мб
image

**Comment by alexhawkins94** [comment]
Switcher 
Заметил, что в наших примерах switcher используется только русская локаль.
При переключении на англ наши примеры не переводятся на другой язык.

**Comment by alexhawkins94** [comment]
Обилие warnings в консоли на странице примеров Таблиц
Default value for val was not provided in AbstractPrizmControl, error in Prizm UI
Angular: Replacing the `context` object of an `EmbeddedViewRef` is deprecated.

image

**Comment by alexhawkins94** [comment]
Table
При входе в режим редактирования ячейки с помощью двойного клика по ячейке в ней сразу же выделяется текст - неудобно.
chrome_6Y3y8uJI37

**Comment by alexhawkins94** [comment]
Table
Sticky bottom - зафиксированная нижняя строка имеет ховер. 
Вроде такого быть не должно и она должна быть аналогично строке заголовков?
image

**Comment by alexhawkins94** [comment]
Tooltip
При изменении tooltip direction не ппроисходит изменение направления отрисовки тултипа.
Задано bottom, тултип по прежнему отображается справа
chrome_hsr5MkqOR3

**Comment by alexhawkins94** [comment]
tabs
при состоянии hover+focus у contained tab обрезается рамка фокуса справа
image

**Comment by alexhawkins94** [comment]
Input carousel - пользовательское значение
Пора бы починить
https://github.com/zyfra/Prizm/issues/1999

File: libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts [hunk 1]
```diff
@@ -279,10 +279,13 @@ export class PrizmInputMultiSelectComponent<T> extends PrizmInputNgControl<T[]>
     this.selectedItems$ = this.value$.pipe(debounceTime(0), startWith(this.value)).pipe(
       switchMap(() => {
         const selectedItems = this.value;
+
         return this.items$.pipe(
           map(items => {
-            return selectedItems.filter(selectedItem => {
-              return items.find(item => this.identityMatcher(selectedItem, this.transformer(item)));
+            return items.filter(selectedItem => {
```
Comment (ickisIckis) [inline libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: @ZurabDev, а почему не устроило первое решение, взятое с  v4?

Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: не работало для трансформеров и объектов

**Comment by alexhawkins94** [comment]
InputDate 
native date - можно ввести значение за пределами min max
image

File: libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts (hunk 1 above)
Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: выше саша описал баги

**Comment by alexhawkins94** [comment]
Input date
Provider for Native Date and ISO or UTC date strings
можно ввести значение за пределами min max
image

**Comment by ickisIckis** [comment]
> InputDate native date - можно ввести значение за пределами min max image

это и следующее  - отстуствующий ранее функционал, решается тут:
https://github.com/zyfra/Prizm/pull/1995

File: libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts (hunk 1 above)
Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: Согласен на счет одного подхода
Тут решение уже не рабочие есть изменение
С 4.6 будет единый подход так сравнение с 4.4 не актуально для 5.1

Comment (ickisIckis) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: Гм, в момент коммита все было очень даже неплохо. 
А какие именно зависимости поменялись?

По 4.6 - надо тогда сразу делать туда коммит, иначе потеряется

Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: делать синхронизацию лучше только после прохождение проверок
если коротоко: решение тут не рабочее так как selectedItems не обладает полным объектом, полные объект со всеми данными до трансформации быть может только в items 

Из за чего появились баги которые выявил Саша

Comment (ickisIckis) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: this.selectedItemsChips$ = combineLatest([this.items$, this.selectedItems$]) - вот тут ты как раз берешь items на v4, как будто бы поменялось только место, где мы вызываем identity mathcer

Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: именно, в я перенес в this.selectedItems$ что он обладал полным набором данных и не приходились обновлять его зависимостях

Comment (ickisIckis) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: Да, минус один лишний вызов функции, все поняла, спасибо.

Надо не забыть в 4.6, отмечу там в комментах

Comment (ZurabDev) [reply libs/components/src/lib/components/dropdowns/multi-select/input-multi-select.component.ts:285]: Спасибо)

**Comment by alexhawkins94** [comment]
Cron 
При переходе в live demo крона отрабатывает событие изменения вкладки Event: selectedChange , Type: PrizmCronTabItem , Data: "month"

Событие отрабатывает два раза

**Comment by alexhawkins94** [comment]
Cron
Продолжение прошлого замечания.
При переключении между вкладками событие отрабатывает дважды Event: selectedChange , Type: PrizmCronTabItem , Data: "day"
chrome_obE4vlO9Mq

**Comment by alexhawkins94** [comment]
TreeMultiSelect
Ошибка в консоли при выполнении поискового запроса.
image

**Comment by alexhawkins94** [comment]
> Tooltip, hint, confirm popup
> 
> При изменении tooltip direction не ппроисходит изменение направления отрисовки тултипа. Задано bottom, тултип по прежнему отображается справа chrome_hsr5MkqOR3
> 
> Аналогично для hint image
> 
> Аналогично для confirm popup image

confirm popup - свойство direction по прежнему не меняет направление отрисовки поп-апа

**Comment by alexhawkins94** [comment]
> По поводу #2093
> 
> При увеличении ширины инпута чипсы не разворачиваются автоматически из-под многоточия. Пересчет и отрисовка свернутых чипсов происходит когда пользователь совершает клик по инпуту.

Наблюдаю ПЛАВАЮЩУЮ проблему при уменьшении ширины инпута- чипсы выходят за пределы.
Воспроизвелось пару раз.
chrome_evuRtgZfJZ

**Comment by alexhawkins94** [comment]
> Обилие warnings в консоли на странице примеров Таблиц, Toogle
> 
> Default value for val was not provided in AbstractPrizmControl, error in Prizm UI Angular: Replacing the `context` object of an `EmbeddedViewRef` is deprecated.
> 
> image

будет исправлено в 5.2 - embedded view
https://github.com/zyfra/Prizm/issues/1506

File: libs/components/src/lib/directives/confirm-popup/confirm-popup.directive.ts [hunk 2]
```diff
@@ -26,174 +26,72 @@ import { PrizmHintDirective } from '../hint/hint.directive';
...
+        'prizmHintHost: prizmConfirmPopupHost',
+        'prizmHintCanShow: prizmConfirmPopupCanShow',
+        'prizmHintContext: prizmConfirmPopupContext',
+      ],
+      outputs: ['prizmHintShowed: prizmConfirmPopupShowed'],
+    },
   ],
   exportAs: 'prizmConfirmPopup',
 })
-export class PrizmConfirmPopupDirective<
```
Comment (ickisIckis) [inline libs/components/src/lib/directives/confirm-popup/confirm-popup.directive.ts:32]: нужно добавить инофрмацию о рефакторинге директивы в описание и changelog

File: libs/components/src/lib/directives/confirm-popup/util.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  };
+
+  return result;
+}
+
+function safeUpdateButtonsWithDefaultStyles(options: PrizmConfirmPopupContext): void {
+  const supportButton = generateButton(
+    options,
+    options.supportButton as string,
+    'Выйти без сохранения',
```
Comment (ickisIckis) [inline libs/components/src/lib/directives/confirm-popup/util.ts:67]: текстовая статика в коде на русском, надо это как-то поправить, тк тербования по локализации

Comment (ZurabDev) [reply libs/components/src/lib/directives/confirm-popup/util.ts:67]: Да это надо. создать задачу общую что пройтись по всем компонентам таким и исправить, в рамках этого релиза таких планов нет, и по регламенту добавить не можем

File: libs/components/src/lib/directives/confirm-popup/confirm-popup.directive.ts (hunk 2 above)
Comment (ZurabDev) [reply libs/components/src/lib/directives/confirm-popup/confirm-popup.directive.ts:32]: спс, готово

**Comment by ickisIckis** [comment]
> date time range руками можно задать значение времени за пределами мин-макс. в value такое значение не прокидывается, однако отображается в инпуте. chrome_NaA9yo087g chrome_wT9zKYCqbm

проблема возникла не в результате текущих задач, скорее всего связана с маской, посмотреть #488
ранее не было условий для выявления этой проблемы

File: libs/components/src/lib/directives/confirm-popup/util.ts (hunk 3 above)
Comment (ickisIckis) [reply libs/components/src/lib/directives/confirm-popup/util.ts:67]: @alexhawkins94

**Review by alexhawkins94 — APPROVED** [review]

**Pull request merged**